packages/markers/src/browser/problem/problem-decorator.ts

```diff
@@ -168,3 +171,131 @@ export namespace ProblemDecorator {
         (left.data.severity || Number.MAX_SAFE_INTEGER) - (right.data.severity || Number.MAX_SAFE_INTEGER);
 
 }
+
+@injectable()
+export class ProblemTabBarDecorator implements TabBarDecorator {
+
+    readonly id = 'theia-problem-tabbar-decorator';
+
+    protected emitter: Emitter<(titles: Title<Widget>[]) => void>;
+
+    @inject(ProblemManager)
+    protected readonly problemManager: ProblemManager;
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @postConstruct()
+    protected init(): void {
+        this.emitter = new Emitter();
+        this.problemManager.onDidChangeMarkers(() => this.fireDidChangeDecorations((titles: Title<Widget>[]) => this.collectDecorators(titles)));
+    }
+
+    decorate(titles: Title<Widget>[]): Map<string, TreeDecoration.Data> {
+        return this.collectDecorators(titles);
+    }
+
+    get onDidChangeDecorations(): Event<(titles: Title<Widget>[]) => void> {
+        return this.emitter.event;
+    }
+
+    protected fireDidChangeDecorations(event: (titles: Title<Widget>[]) => Map<string, TabBarDecoration.Data>): void {
+        this.emitter.fire(event);
+    }
+
+    /**
+     * Collect decorators for the tabs.
+     * @returns {Map<string, TabBarDecoration.Data>} A map from the tab URI to the tab decoration data.
+     */
+    protected collectDecorators(titles: Title<Widget>[]): Map<string, TabBarDecoration.Data> {
+        const result: Map<string, Marker<Diagnostic>> = new Map();
+        const markers = this.groupMarkersByURI(this.collectMarkers());
+        for (const title of titles) {
+            const relativeUri = title.caption;
+            const root = this.workspaceService.tryGetRoots()[0];
+            const path = root.uri + '/' + relativeUri;
```

Comment:
we should use `URI` and `Path` apis to manipulate uris and paths, not string operations

Reply:
added https://github.com/theia-ide/theia/wiki/Coding-Guidelines#uri-no-string-manipulations